src/Dianoga/Invokers/MediaCacheAsync/OptimizingMediaCache.cs

```diff
@@ -50,66 +53,80 @@ public override bool AddStream(Media media, MediaOptions options, MediaStream st
+					// (which will dispose of originalStream)
 					var backupStream = new MemoryStream();
-					copyStream.CopyTo(backupStream);
+					originalStream.CopyTo(backupStream);
 					backupStream.Seek(0, SeekOrigin.Begin);
 
-					var backupMediaStream = new MediaStream(backupStream, stream.Extension, stream.MediaItem);
-
-					// switch to the right site context (see above)
-					using (new SiteContextSwitcher(currentSite))
-					{
-						MediaCacheRecord cacheRecord = null;
-
-						var optimizedStream = _optimizer.Process(copiedMediaStream, options);
-
-						if (optimizedStream == null)
-						{
-							Log.Info($"Dianoga: {mediaPath} cannot be optimized due to media type or path exclusion", this);
-							cacheRecord = CreateCacheRecord(media, options, backupMediaStream);
-						}
-
-						if (cacheRecord == null)
-						{
-							cacheRecord = CreateCacheRecord(media, options, optimizedStream);
-							backupMediaStream.Dispose();
-						}
-
-						AddToActiveList(cacheRecord);
-
-						cacheRecord.Persist();
-
-						RemoveFromActiveList(cacheRecord);
-					}
-				}
-				catch (Exception ex)
-				{
-					// this runs in a background thread, and an exception here would cause IIS to terminate the app pool. Bad! So we catch/log, just in case.
-					Log.Error($"Dianoga: Exception occurred on the background thread when optimizing: {mediaPath}", ex, this);
+					backupMediaStream = new MediaStream(backupStream, mediaItem.Extension, mediaItem);
 				}
-			});
 
-			return true;
-		}
+				// switch to the right site context (see above)
+				using (new SiteContextSwitcher(currentSite))
+                {
+                    MediaCacheRecord cacheRecord = null;
+
+                    optimizedMediaStream = _optimizer.Process(originalMediaStream, options);
+
+                    if (optimizedMediaStream == null)
+                    {
+                        Log.Info($"Dianoga: {mediaItem.MediaPath} cannot be optimized due to media type or path exclusion", this);
+                        cacheRecord = CreateCacheRecord(media, options, backupMediaStream);
+                    }
+
+                    if (cacheRecord == null)
+                    {
+	                    cacheRecord = CreateCacheRecord(media, options, optimizedMediaStream);
+                    }
+
+                    AddToActiveList(cacheRecord);
+
+                    cacheRecord.Persist();
+
+                    RemoveFromActiveList(cacheRecord);
+                }
+            }
+            catch (Exception ex)
+            {
+	            // this runs in a background thread, and an exception here would cause IIS to terminate the app pool. Bad! So we catch/log, just in case.
+				Log.Error($"Dianoga: Exception occurred on the background thread when optimizing: {mediaItem.MediaPath}", ex, this);
+            }
+            finally
+            {
+				// release resources used by the optimization task
+                originalMediaStream?.Dispose();
```

Comment:
Can you test that in the case where the optimisation failed for some reason the original stream is returned and original image is rendered? As a test you could use the corrupted image from here https://github.com/kamsar/Dianoga/issues/68

Reply:
Isn't that already done in ShouldNotSquishCorruptedJpegLossy in JpegOptimOptimizerTests? If the OptimizingMediaCache.cs needs to be tested we need to add fakes for the Sitecore objects as well. 
If still needed I'll try to have a look into it.

Reply:
Yeah good point. Would be nice to have tests for OptimizingMediaCache if you are volunteering :). I'll give this PR a test soon.